### PR TITLE
feature/frozen-features

### PR DESCRIPTION
**Pull request by gnusi**
Compile time server feature lookup

### CODE REVIEW

File: lib/ApplicationFeatures/ApplicationServer.h [hunk 1]
```diff
@@ -227,46 +225,42 @@ class ApplicationServer {
...
-   bool hasFeature(std::type_index type) const noexcept {
-     return (_features.find(type) != _features.end());
-   }
-
    // checks for the existence of a feature. will not throw when used for
    // a non-existing feature
    template <typename Type, typename std::enable_if<std::is_base_of<ApplicationFeature, Type>::value, int>::type = 0>
    bool hasFeature() const noexcept {
-     return hasFeature(std::type_index(typeid(Type)));
+     return nullptr == _features[Feature<Type>::Index];
```
Comment (jsteemann) [inline lib/ApplicationFeatures/ApplicationServer.h:244]:
```suggestion
     return nullptr != _features[Feature<Type>::Index];
```

File: lib/ApplicationFeatures/ApplicationServer.h [hunk 2]
```diff
@@ -29,11 +29,11 @@
 #include <memory>
 #include <string>
 #include <type_traits>
-#include <typeindex>
```
Comment (jsteemann) [inline lib/ApplicationFeatures/ApplicationServer.h:32]: I think we should also include `<array>` here, as this header is now using it.

**Comment by danielhlarkin** [comment]
I don't necessarily object to a compile-time feature map, but I do have some serious concerns about the current implementation centralizing all the types from server, shell, enterprise, etc. into `lib` via forward declarations. Is it possible instead to build some template class in lib that gets instantiated where the server is actually instantiated, e.g. in `arangod/RestServer/arangod.cpp`, `arangosh/Restore/arangorestore.cpp`, etc.?

**Comment by gnusi** [comment]
@danielhlarkin absolutely, but all types have to be known per binary beforehand.

**Comment by gnusi** [comment]
array vs unordered_map lookup
https://quick-bench.com/q/6S8N9fmH2dHOEICiRW24R4vl37c

**Comment by gnusi** [comment]
superseded with #15610

**Pull request closed without merging**